### PR TITLE
NixOS module for kupo

### PR DESCRIPTION
**Pull request by euonymos**

### CODE REVIEW

File: kupo-nixos-module.nix [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+    };
+
+    nodeSocket = mkOption {
+      description = "Path to cardano-node IPC socket.";
+      type = str;
+    };
+
+    nodeConfig = mkOption {
+      description = "Path to cardano-node config.json file.";
+      type = str;
```
Comment (brainrake) [inline kupo-nixos-module.nix]: path

File: kupo-nixos-module.nix [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+    group = mkOption {
+      description = "Group to run kupo service as.";
+      type = types.str;
+      default = "kupo";
+    };
+
+    nodeSocket = mkOption {
+      description = "Path to cardano-node IPC socket.";
+      type = str;
```
Comment (brainrake) [inline kupo-nixos-module.nix]: path

File: kupo-nixos-module.nix [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+      enable = true;
+      after = [ "cardano-node.service" ];
+      wantedBy = [ "multi-user.target" ];
+
+      script = escapeShellArgs (concatLists [
+        [ "${cfg.package}/bin/kupo" ]
+        [ "--node-socket" cfg.nodeSocket ]
+        [ "--node-config" cfg.nodeConfig ]
+        [ "--host" cfg.host ]
+        # [ "--workdir" "/tmp" ]
```
Comment (brainrake) [inline kupo-nixos-module.nix]: add an option or remove this (can still be added via `extraArgs`). best if systemd tmpdir support is used, see PrivateTmp

File: kupo-nixos-module.nix [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+        [ "--in-memory" ]
+        [ "--match" "*/*" ]
+        [ "--since" "origin" ]
+        cfg.extraArgs
+      ]);
+
+      serviceConfig = {
+        User = cfg.user;
+        Group = cfg.group;
+        # Security
```
Comment (brainrake) [inline kupo-nixos-module.nix]: please run `systemd-analyze security` and try to improve based on suggestions there, unless the service breaks.

**Review by brainrake — CHANGES_REQUESTED** [review]
looking good, just a few minor changes needed

**Comment by brainrake** [comment]
why do we have a fork? can we upstream?

**Comment by euonymos** [comment]
> why do we have a fork? can we upstream?

I got the impression that Matthias is not really fond of Nix... 
https://github.com/CardanoSolutions/kupo/pull/55

**Comment by euonymos** [comment]
> why do we have a fork? can we upstream?

BTW what's the best practice if we need a NixOS module, but upstream repo doesn't want to accept any Nix code?

**Comment by brainrake** [comment]
i reckon, make a flake repo with the nixos module and reference the project repo.

**Comment by brainrake** [comment]
Implemented in a separate repo.
https://github.com/mlabs-haskell/kupo-nixos

**Pull request closed without merging**